### ISSUE TITLE
Kubernetes docker image is broken

### ISSUE THREAD

**Issue by germanoeich** [tailscale/tailscale#4537]
### What is the issue?

This issue here is both a feature request of sorts and a bug report.

tl;dr: The kubernetes deployment experience is terrible.

First off the bug:

I followed the tutorial for kubernetes for setting a subnet-router, and no matter what I try, I always hit this:

```
/run.sh: line 4: 
: not found
/run.sh: line 6: 
: not found
/run.sh: line 8: 
: not found
/run.sh: line 15: 
: not found
/run.sh: set: line 16: illegal option -
```

I tried fiddling around with the shell script to no avail. This is a kubernetes cluster running on top of Rancher Kubernetes Engine, running k8s 1.21 on ubuntu 20.04.

Now for the Kubernetes experience part:

The user experience for tailscale is out of this world, but I can't say the same for k8s. I spent literally 3 hours messing around with this to end up deploying in a VM. Here are the pain points I felt:

1. Please, provide a pre-built container image with the run.sh script setup or support ENV vars in the go application. It feels like a workaround to have to build your own container and it hurts devops teams, because now we have to setup a pipeline just to keep tailscale images up to date.
2. Using a secret to hold state seems like a wierd choice. Why not use the ephemeral disk? Or even a block storage volume? Why does the agent need to hold data when the control plane is managed? Is this just a tradeoff for faster startups? If so, we should have a choice to opt out.
3. The choice of having a standalone pod yaml is non-standard. In any other projects, you'd see a deployment being used. Standalone pods don't play nice with tools like Rancher and can't be properly updated, redeployed, etc through those tools.
4. Namespace support is lacking

As for the documentation around kubernetes, it's very basic and confusing. It's a step by step thing that uses make to obscure the steps. Can we get proper documentation? Outlining what USERSPACE does, what NET_ capabilities our containers must have and in what context we need to run priviledged containers, etc?

### Steps to reproduce

using the docs/k8s files:

```
<Setup variables, in this case I did: SA_NAME, IMAGE_TAG (Pointing to a github registry), KUBE_SECRET, ROUTES>
<Create the secret holding AUTH_KEY>
make push
make rbac
make subnet-router
```

Env values used:
```
SA_NAME=tailscale
IMAGE_TAG=ghcr.io/myorg/tailscale-k8s
KUBE_SECRET=tailscale
ROUTES=10.12.0.0/24
```

### Are there any recent changes that introduced the issue?

This is a new deployment.

### OS

Linux

### OS version

Ubuntu 20.04

### Tailscale version

1.24.0

### Bug report

Could not get the agent to run.

**Comment by lukehsiao** [comment]
I'm experiencing the same pain points. In my case, I'm looking to run the non-userspace sidecar in AWS EKS.

I've also created the docker image, run through the tutorial for the sidecar, but when launching the nginx pod, I see this:

```
❯ kubectl logs nginx ts-sidecar
Starting tailscaled
Running tailscale up
2022/05/04 20:19:57 logtail started
2022/05/04 20:19:57 Program starting: v1.24.2-tdce2409b1, Go 1.18.1-ts710a0d8610: []string{"tailscaled", "--state=kube:tailscale-auth", "--socket=/tmp/tailscaled.sock"}
2022/05/04 20:19:57 LogID: 0ac05a2c564edf28ba684e7a4675fc45098dfc3b40e676cf980cb9de6cb69ba7
2022/05/04 20:19:57 logpolicy: using system state directory "/var/lib/tailscale"
logpolicy.Read /var/lib/tailscale/tailscaled.log.conf: open /var/lib/tailscale/tailscaled.log.conf: no such file or directory
2022/05/04 20:19:57 wgengine.NewUserspaceEngine(tun "tailscale0") ...
2022/05/04 20:19:57 setting link attributes: netlink receive: no such file or directory
2022/05/04 20:19:57 router: disabling tunneled IPv6 due to system IPv6 config: disable_ipv6 is set
2022/05/04 20:19:57 dns: [rc=unknown ret=direct]
2022/05/04 20:19:57 dns: using *dns.directManager
2022/05/04 20:19:57 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[10.0.3.49/32]} v4=true v6=false}
2022/05/04 20:19:57 magicsock: disco key = d:baade28b8026e6bc
2022/05/04 20:19:57 Creating wireguard device...
2022/05/04 20:19:57 Bringing wireguard device up...
2022/05/04 20:19:57 Bringing router up...
2022/05/04 20:19:57 Clearing router settings...
2022/05/04 20:19:57 Starting link monitor...
2022/05/04 20:19:57 Engine created.
2022/05/04 20:19:57 external route: up
2022/05/04 20:19:57 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:254 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5210 Table:254 Mark:4294967295 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2022/05/04 20:19:57 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:253 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5230 Table:253 Mark:4294967295 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2022/05/04 20:19:57 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:0 Protocol:0 Scope:0 Type:7 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5250 Table:0 Mark:4294967295 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2022/05/04 20:19:57 monitor: ip rule deleted: {Family:2 DstLength:0 SrcLength:0 Tos:0 Table:52 Protocol:0 Scope:0 Type:1 Flags:0 Attributes:{Dst:<nil> Src:<nil> Gateway:<nil> OutIface:0 Priority:5270 Table:52 Mark:0 Pref:<nil> Expires:<nil> Metrics:<nil> Multipath:[]}}
2022/05/04 20:19:57 Start
2022/05/04 20:19:57 using backend prefs for "_daemon": Prefs{ra=false dns=false want=true routes=[172.20.0.0/16 10.0.1.0/24 10.0.2.0/24 10.0.3.0/24] snat=true nf=on Persist{lm=, o=, n=[LBNY2] u="myemail@example.com"}}
2022/05/04 20:19:57 Backend: logs: be:0ac05a2c564edf28ba684e7a4675fc45098dfc3b40e676cf980cb9de6cb69ba7 fe:
2022/05/04 20:19:57 control: client.Login(false, 0)
2022/05/04 20:19:57 control: doLogin(regen=false, hasUrl=false)
Error: changing settings via 'tailscale up' requires mentioning all
non-default flags. To proceed, either re-run your command with --reset or
use the command below to explicitly mention the current value of
all non-default settings:

        tailscale up --accept-dns=false --auth-key=tskey-k82xLY4CNTRL-Vc6dYoxXbB5ZhMmBxFc8t --advertise-routes=172.20.0.0/16,10.0.1.0/24,10.0.2.0/24,10.0.3.0/24
```

So, it seems to me that there is also potentially a bug with run.sh itself, in that it appears to be invoking tailscale in an incompatible way.

**Comment by bradfitz** [comment]
cc @maisem

**Comment by lukehsiao** [comment]
Another thing I've found running through the examples is a bug similar to #4414.

<details open><summary>k8s logs</summary>
  
```
❯ kubectl logs subnet-router
+ set -eo pipefail
+ TAILSCALED_ARGS='--state=kube:tailscale-auth --socket=/tmp/tailscaled.sock'
+ '[[' true '==' true ]]
+ '[[' '!' -z  ]]
+ TAILSCALED_ARGS='--state=kube:tailscale-auth --socket=/tmp/tailscaled.sock --tun=userspace-networking'
+ echo 'Starting tailscaled'
+ PID=7
+ UP_ARGS='--accept-dns=false'
+ '[[' '!' -z 172.20.0.0/16,10.0.1.0/24,10.0.2.0/24,10.0.3.0/24 ]]
+ UP_ARGS='--advertise-routes=172.20.0.0/16,10.0.1.0/24,10.0.2.0/24,10.0.3.0/24 --accept-dns=false'
+ + '[[' '!' -ztailscaled tskey-k82xLY4CNTRL-Vc6dYoxXbB5ZhMmBxFc8t '--state=kube:tailscale-auth' ]] '--socket=/tmp/tailscaled.sock'
 '--tun=userspace-networking'
+ UP_ARGS='--authkey=tskey-k82xLY4CNTRL-Vc6dYoxXbB5ZhMmBxFc8t --advertise-routes=172.20.0.0/16,10.0.1.0/24,10.0.2.0/24,10.0.3.0/24 --accept-dns=false'
+ '[[' '!' -z  ]]
+ echo 'Running tailscale up'
+ tailscale '--socket=/tmp/tailscaled.sock' up '--authkey=tskey-k82xLY4CNTRL-Vc6dYoxXbB5ZhMmBxFc8t' '--advertise-routes=172.20.0.0/16,10.0.1.0/24,10.0.2.0/24,10.0.3.0/24' '--accept-dns=false'
Starting tailscaled
Running tailscale up
2022/05/04 23:54:46 logtail started
2022/05/04 23:54:46 Program starting: v1.24.2-tdce2409b1, Go 1.18.1-ts710a0d8610: []string{"tailscaled", "--state=kube:tailscale-auth", "--socket=/tmp/tailscaled.sock", "--tun=userspace-networking"}
2022/05/04 23:54:46 LogID: adcb843904550acdf46e590b0fa324f3730e6fddb5b9eb498218f796d470267a
2022/05/04 23:54:46 filch failed: filch: open /.cache/Tailscale/tailscaled.log1.txt: no such file or directory
2022/05/04 23:54:46 logpolicy: using UserCacheDir, "/.cache/Tailscale"
logpolicy.Read /.cache/Tailscale/tailscaled.log.conf: open /.cache/Tailscale/tailscaled.log.conf: no such file or directory
logpolicy.Config.Save: mkdir /.cache: permission denied
2022/05/04 23:54:46 wgengine.NewUserspaceEngine(tun "userspace-networking") ...
2022/05/04 23:54:46 dns: using dns.noopManager
2022/05/04 23:54:46 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[10.0.3.107/32]} v4=true v6=false}
2022/05/04 23:54:46 magicsock: disco key = d:f949360c0f9ef017
2022/05/04 23:54:46 Creating wireguard device...
2022/05/04 23:54:46 Bringing wireguard device up...
2022/05/04 23:54:46 Bringing router up...
2022/05/04 23:54:46 Clearing router settings...
2022/05/04 23:54:46 Starting link monitor...
2022/05/04 23:54:46 Engine created.
2022/05/04 23:54:46 Start
2022/05/04 23:54:46 using backend prefs for "_daemon": Prefs{ra=false dns=false want=true routes=[172.20.0.0/16 10.0.1.0/24 10.0.2.0/24 10.0.3.0/24] snat=true nf=on Persist{lm=, o=, n=[LBNY2] u="luke.hsiao@numbersstation.ai"}}
2022/05/04 23:54:46 Backend: logs: be:adcb843904550acdf46e590b0fa324f3730e6fddb5b9eb498218f796d470267a fe:
2022/05/04 23:54:46 control: client.Login(false, 0)
2022/05/04 23:54:46 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:46 ipnserver: conn2: connection from userid 1000; connection from non-root user matching daemon has access
2022/05/04 23:54:46 Start
2022/05/04 23:54:46 control: client.Shutdown()
2022/05/04 23:54:46 control: client.Shutdown: inSendStatus=0
2022/05/04 23:54:46 control: mapRoutine: quit
2022/05/04 23:54:46 control: Client.Shutdown done.
2022/05/04 23:54:46 using backend prefs for "_daemon": Prefs{ra=false dns=false want=true routes=[172.20.0.0/16 10.0.1.0/24 10.0.2.0/24 10.0.3.0/24] snat=true nf=on Persist{lm=, o=, n=[LBNY2] u="luke.hsiao@numbersstation.ai"}}
2022/05/04 23:54:46 Backend: logs: be:adcb843904550acdf46e590b0fa324f3730e6fddb5b9eb498218f796d470267a fe:
2022/05/04 23:54:46 control: client.Login(false, 0)
2022/05/04 23:54:46 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:47 control: control server key from https://controlplane.tailscale.com: ts2021=[fSeS+], legacy=[nlFWp]
2022/05/04 23:54:47 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:55 Received error: register request: http 500: internal server error
2022/05/04 23:54:55 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:55 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:55 Received error: register request: http 500: internal server error
2022/05/04 23:54:55 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:55 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:56 Received error: register request: http 500: internal server error
2022/05/04 23:54:56 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:56 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:56 Received error: register request: http 500: internal server error
2022/05/04 23:54:56 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:56 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:56 Received error: register request: http 500: internal server error
2022/05/04 23:54:56 [RATELIMIT] format("Received error: %v")
2022/05/04 23:54:57 control: doLogin(regen=false, hasUrl=false)
2022/05/04 23:54:57 [RATELIMIT] format("control: doLogin(regen=%v, hasUrl=%v)")
2022/05/04 23:54:57 control: RegisterReq: onode= node=[LBNY2] fup=false
2022/05/04 23:54:57 [RATELIMIT] format("control: RegisterReq: onode=%v node=%v fup=%v")
2022/05/04 23:55:03 [RATELIMIT] format("control: RegisterReq: onode=%v node=%v fup=%v") (5 dropped)
2022/05/04 23:55:03 control: RegisterReq: onode= node=[LBNY2] fup=false
```

</details>

It seems the solution in the k8s case, which stores the state in `kube:<secret>`, rather than `/var/lib/tailscale` is to clear that state. The more reliable approach for these inherently ephemeral nodes seems to be to use `mem:` instead of `kube:<secret>` to me. Is there any downsides to that approach? Why not make that the default for `run.sh`?

**Comment by shadycuz** [comment]
I also found it very painful. From the documentation https://tailscale.com/kb/1185/kubernetes/, I assumed I could just use the tailscale published image.

> Tailscale has a published Docker image that Tailscale manages and builds from source. It’s available in [Docker Hub](https://hub.docker.com/r/tailscale/tailscale) and [GitHub Packages](https://github.com/orgs/tailscale/packages/container/package/tailscale), just run:
> 
> docker pull tailscale/tailscale:latest
> 
> or
> 
> docker pull ghcr.io/tailscale/tailscale:latest


When it didn't work, I found this issue. Now I understand that I have to build the docker sidecar image my self.

**Comment by MrSaints** [comment]
I made the same mistake of using the published image instead of building a new one. I think it is pretty misleading to include the `docker pull` section.This should be done automatically on a build anyway.

**Comment by shadycuz** [comment]
I finally got it running once I built my own container but I while it does work, it behaves weird. I will open an new issue for this:

```
/ # tailscale status
failed to connect to local tailscaled (which appears to be running as tailscaled, pid 9). Got error: Failed to connect to local Tailscale daemon for /localapi/v0/status; not running? Error: dial unix /var/run/tailscale/tailscaled.sock: connect: no such file or directory
```

**Comment by shadycuz** [comment]
It seems they have now added K8 support to the tailscale docker image at the root of the repo. See #4799 

This means all docker tags at `1.26.X` and up (including latest), will have support for K8's without building your own container. 

I haven't tested this but will soon. 🚀

**Comment by shadycuz** [comment]
It also appears they changes the values of the environment variables. So if your container used `KUBE_SECRET`, `USERSPACE` or `AUTH_KEY` you should pre-pend `TS_` to them. 

Like this:
```yaml
env:
- name: TS_KUBE_SECRET
  value: "tailscale-auth"
- name: TS_USERSPACE
  value: "false"
- name: TS_AUTH_KEY
  valueFrom:
    secretKeyRef:
      name: tailscale-auth
      key: AUTH_KEYthe state in a k8s secret
```

**Comment by noamApps** [comment]
> It also appears they changed the values of the environment variables. So if your container used `KUBE_SECRET`, `USERSPACE` or `AUTH_KEY` you should pre-pend `TS_` to them.
> 
> Like this:
> 
> ```yaml
> env:
> - name: TS_KUBE_SECRET
>   value: "tailscale-auth"
> - name: TS_USERSPACE
>   value: "false"
> - name: TS_AUTH_KEY
>   valueFrom:
>     secretKeyRef:
>       name: tailscale-auth
>       key: AUTH_KEY
> ```

we have also upgraded to Tailscale 1.26 using the Tailscale-hosted image and had to add the TS_ prefix to make it work.
the docs needs to be updated as they mention the old environment variable ( for example https://tailscale.com/kb/1185/kubernetes/ )

**Comment by DentonGentry** [comment]
The doc has been updated to add TS_ prefixes. https://tailscale.com/kb/1185/kubernetes/

**Comment by mayakacz** [comment]
> I made the same mistake of using the published image instead of building a new one.

This was addressed in June: https://github.com/tailscale/tailscale/pull/4799

> it seems to me that there is also potentially a bug with run.sh itself, in that it appears to be invoking tailscale in an incompatible way.

@maisem I think you said this should be updated?

Is there anything else unaddressed by this bug?
